Proposed Fix for Issue #340, don't allow nested IFs on same w:p or w:…

dseiler opened the pull request:
Proposed Fix for Issue #340, don't allow nested IFs on same w:p or w:r tag.
Added some code to the processTemplate.ts file to throw an exception in the processForIf function if an IF statement is nested within another IF on the same w:p or w:r tag. 
I had to add two map variables to the context for keeping track the corrent w:p or w:r node.
I also added three tests. Those tests run into an infinite loop if my fix is not applied. All the other tests run correctly. Please verify and if ok add to the next version of docx-templates. Thanks!

jjhbw commented on package-lock.json:
The project uses yarn, can you remove package-lock.json?

jjhbw commented on package.json:
```diff
@@ -50,7 +50,7 @@
     "@types/qrcode": "1.5.0",
     "@types/sax": "^1.2.4",
-    "@typescript-eslint/eslint-plugin": "^6.5.0",
+    "@typescript-eslint/eslint-plugin": "^7.0.1",
```
Can you revert this update? I prefer updating dev deps in separate commits, outside of feature MRs.

jjhbw commented on src/processTemplate.ts:
```diff
@@ -167,6 +170,45 @@ const debugPrintNode = (node: Node) =>
         }
   );
 
+// NEW DSE
```
Please just remove the previous version of this function instead of commenting it out, then it is also easier to see for me in the diff exactly what you changed.

jjhbw commented on src/processTemplate.ts:
```diff
@@ -386,6 +428,22 @@ export async function walkTemplate(
       // Clone input node and append to output tree
       const newNode: Node = cloneNodeWithoutChildren(nodeIn);
+
+      // NEW DSE if the new node is p or tr node, add it to the map to check whether two for/ifs are nested on the same line
```
Same here; is this code unused? If so please remove the commented out code.

jjhbw commented on yarn.lock:
Same here: can you revert any changes to `yarn.lock` unless the changes to the deps are crucial for your MR (and I don't think they are, right?).

jjhbw commented on src/types.ts:
```diff
@@ -168,6 +168,10 @@ export type Context = {
   jsSandbox?: SandBox;
   textRunPropsNode?: NonTextNode;
+
+  // NEW DSE
```
Can you remove the 'NEW DSE' comments everywhere? Seems like redundant information. If we want to know why these two `Map`s exist in the future, we can just use `git blame` and see your awesome work.

jjhbw reviewed (changes requested):
Awesome work, thanks! I added a few small nitpicks. Would be great if you could change those before I run CI.

dseiler commented:
Thanks for your feedback. Highly appreciated. I updated my pull request with your comments. I also added a new "emergency exit" to the main while loop in case something runs out of control. I.e. if we run into an endless loop we should throw and exception. Let me know if this makes sense. Thanks!

jjhbw commented:
Reverted the `yarn.lock` change and squash-merged this MR offline, making sure to retain your authorship info. See f069eb22469a296db555faafbb164fa7b68c7f7f

Thanks for the contribution, and apologies for my late reactions.

xymfs commented:
I have the same problem. After update #340 , a lot of my documents cannot be generated!
like:
```
{{IF $i.result.includes('xxx')}}
{{FOR user IN $i.users}}
{{IF $user.status === true}}
{{ $user.username }}：{{$user.age}}（{{userpoint.sex}}）
{{END-IF}}
{{END-FOR user}}
{{END-IF}}
```
